README
NAME

    Data::Object::Space - Namespace Class

ABSTRACT

    Namespace Class for Perl 5

SYNOPSIS

      package main;
    
      use Data::Object::Space;
    
      my $space = Data::Object::Space->new('foo/bar');

DESCRIPTION

    This package provides methods for parsing and manipulating package
    namespaces.

INHERITS

    This package inherits behaviors from:

    Data::Object::Name

LIBRARIES

    This package uses type constraints from:

    Types::Standard

METHODS

    This package implements the following methods:

 append

      append(Str @args) : Object

    The append method modifies the object by appending to the package
    namespace parts.

    append example #1

        # given: synopsis
      
        $space->append('baz');
      
        # 'Foo/Bar/Baz'

    append example #2

        # given: synopsis
      
        $space->append('baz', 'bax');
      
        # $space->package;
      
        # 'Foo/Bar/Baz/Bax'

 array

      array(Str $arg1) : ArrayRef

    The array method returns the value for the given package array variable
    name.

    array example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our @handler = 'start';
      
        package main;
      
        $space->array('handler')
      
        # ['start']

 arrays

      arrays() : ArrayRef

    The arrays method searches the package namespace for arrays and returns
    their names.

    arrays example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our @handler = 'start';
        our @initial = ('next', 'prev');
      
        package main;
      
        $space->arrays
      
        # ['handler', 'initial']

 authority

      authority() : Maybe[Str]

    The authority method returns the AUTHORITY declared on the target
    package, if any.

    authority example #1

        package Foo::Boo;
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/boo');
      
        $space->authority
      
        # undef

    authority example #2

        package Foo::Boo;
      
        our $AUTHORITY = 'cpan:AWNCORP';
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/boo');
      
        $space->authority
      
        # 'cpan:AWNCORP'

 base

      base() : Str

    The base method returns the last segment of the package namespace
    parts.

    base example #1

        # given: synopsis
      
        $space->base
      
        # Bar

 bless

      bless(Any $arg1 = {}) : Object

    The bless method blesses the given value into the package namespace and
    returns an object. If no value is given, an empty hashref is used.

    bless example #1

        # given: synopsis
      
        package Foo::Bar;
      
        sub import;
      
        package main;
      
        $space->bless
      
        # bless({}, 'Foo::Bar')

    bless example #2

        # given: synopsis
      
        package Foo::Bar;
      
        sub import;
      
        package main;
      
        $space->bless({okay => 1})
      
        # bless({okay => 1}, 'Foo::Bar')

 build

      build(Any @args) : Object

    The build method attempts to call new on the package namespace and if
    successful returns the resulting object.

    build example #1

        package Foo::Bar::Baz;
      
        sub new {
          bless {}, $_[0]
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/bar/baz');
      
        $space->build
      
        # bless({}, 'Foo::Bar::Baz')

    build example #2

        package Foo::Bar::Bax;
      
        sub new {
          bless $_[1], $_[0]
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/bar/bax');
      
        $space->build({okay => 1})
      
        # bless({okay => 1}, 'Foo::Bar::Bax')

 call

      call(Any @args) : Any

    The call method attempts to call the given subroutine on the package
    namespace and if successful returns the resulting value.

    call example #1

        # given: synopsis
      
        package Foo;
      
        sub import;
      
        sub start {
          'started'
        }
      
        package main;
      
        use Data::Object::Space;
      
        $space = Data::Object::Space->new('foo');
      
        $space->call('start')
      
        # started

    call example #2

        # given: synopsis
      
        package Zoo;
      
        sub import;
      
        sub AUTOLOAD {
          bless {};
        }
      
        sub DESTROY {
          ; # noop
        }
      
        package main;
      
        use Data::Object::Space;
      
        $space = Data::Object::Space->new('zoo');
      
        $space->call('start')
      
        # bless({}, 'Zoo')

 child

      child(Str $arg1) : Object

    The child method returns a new Data::Object::Space object for the child
    package namespace.

    child example #1

        # given: synopsis
      
        $space->child('baz');
      
        # $space->package;
      
        # Foo::Bar::Baz

 children

      children() : ArrayRef[Object]

    The children method searches %INC and @INC and retuns a list of
    Data::Object::Space objects for each child namespace found (one level
    deep).

    children example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('c_p_a_n');
      
        $space->children
      
        # [
        #   'CPAN/Author',
        #   'CPAN/Bundle',
        #   'CPAN/CacheMgr',
        #   ...
        # ]

 cop

      cop(Any @args) : CodeRef

    The cop method attempts to curry the given subroutine on the package
    namespace and if successful returns a closure.

    cop example #1

        # given: synopsis
      
        package Foo::Bar;
      
        sub import;
      
        sub handler {
          [@_]
        }
      
        package main;
      
        use Data::Object::Space;
      
        $space = Data::Object::Space->new('foo/bar');
      
        $space->cop('handler', $space->bless)
      
        # sub { Foo::Bar::handler(..., @_) }

 destroy

      destroy() : Object

    The destroy method attempts to wipe out a namespace and also remove it
    and its children from %INC. NOTE: This can cause catastrophic failures
    if used incorrectly.

    destroy example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('data/dumper');
      
        $space->load; # Data/Dumper
      
        $space->destroy;

 eval

      eval(Str @args) : Any

    The eval method takes a list of strings and evaluates them under the
    namespace represented by the instance.

    eval example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo');
      
        $space->eval('our $VERSION = 0.01');

 functions

      functions() : ArrayRef

    The functions method searches the package namespace for functions and
    returns their names.

    functions example #1

        package Foo::Functions;
      
        use routines;
      
        fun start() {
          1
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/functions');
      
        $space->functions
      
        # ['start']

 hash

      hash(Str $arg1) : HashRef

    The hash method returns the value for the given package hash variable
    name.

    hash example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our %settings = (
          active => 1
        );
      
        package main;
      
        $space->hash('settings')
      
        # {active => 1}

 hashes

      hashes() : ArrayRef

    The hashes method searches the package namespace for hashes and returns
    their names.

    hashes example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our %defaults = (
          active => 0
        );
      
        our %settings = (
          active => 1
        );
      
        package main;
      
        $space->hashes
      
        # ['defaults', 'settings']

 id

      id() : Str

    The id method returns the fully-qualified package name as a label.

    id example #1

        # given: synopsis
      
        $space->id
      
        # Foo_Bar

 included

      included() : Str

    The included method returns the path of the namespace if it exists in
    %INC.

    included example #1

        package main;
      
        my $space = Data::Object::Space->new('Data/Object/Space');
      
        $space->included;
      
        # lib/Data/Object/Space.pm

 inherits

      inherits() : ArrayRef

    The inherits method returns the list of superclasses the target package
    is derived from.

    inherits example #1

        package Bar;
      
        package main;
      
        my $space = Data::Object::Space->new('bar');
      
        $space->inherits
      
        # []

    inherits example #2

        package Foo;
      
        package Bar;
      
        use base 'Foo';
      
        package main;
      
        my $space = Data::Object::Space->new('bar');
      
        $space->inherits
      
        # ['Foo']

 init

      init() : Str

    The init method ensures that the package namespace is loaded and,
    whether created in-memory or on-disk, is flagged as being loaded and
    loadable.

    init example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('kit');
      
        $space->init
      
        # Kit

 inject

      inject(Str $name, Maybe[CodeRef] $coderef) : Any

    The inject method monkey-patches the package namespace, installing a
    named subroutine into the package which can then be called normally,
    returning the fully-qualified subroutine name.

    inject example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('kit');
      
        $space->inject('build', sub { 'finished' });
      
        # *Kit::build

 load

      load() : Str

    The load method checks whether the package namespace is already loaded
    and if not attempts to load the package. If the package is not loaded
    and is not loadable, this method will throw an exception using confess.
    If the package is loadable, this method returns truthy with the package
    name. As a workaround for packages that only exist in-memory, if the
    package contains a new, with, meta, or import routine it will be
    recognized as having been loaded.

    load example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('c_p_a_n');
      
        $space->load
      
        # CPAN

 loaded

      loaded() : Int

    The loaded method checks whether the package namespace is already
    loaded returns truthy or falsy.

    loaded example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('data/dumper');
      
        $space->loaded;
      
        # 0

    loaded example #2

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('data/dumper');
      
        $space->load;
      
        $space->loaded;
      
        # 1

 locate

      locate() : Str

    The locate method checks whether the package namespace is available in
    @INC, i.e. on disk. This method returns the file if found or an empty
    string.

    locate example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('brianne_spinka');
      
        $space->locate;
      
        # ''

    locate example #2

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('data/dumper');
      
        $space->locate;
      
        # /path/to/Data/Dumper.pm

 methods

      methods() : ArrayRef

    The methods method searches the package namespace for methods and
    returns their names.

    methods example #1

        package Foo::Methods;
      
        use routines;
      
        method start() {
          1
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/methods');
      
        $space->methods
      
        # ['start']

 name

      name() : Str

    The name method returns the fully-qualified package name.

    name example #1

        # given: synopsis
      
        $space->name
      
        # Foo::Bar

 parent

      parent() : Object

    The parent method returns a new Data::Object::Space object for the
    parent package namespace.

    parent example #1

        # given: synopsis
      
        $space->parent;
      
        # $space->package;
      
        # Foo

 parse

      parse() : ArrayRef

    The parse method parses the string argument and returns an arrayref of
    package namespace segments (parts).

    parse example #1

        my $space = Data::Object::Space->new('Foo::Bar');
      
        $space->parse;
      
        # ['Foo', 'Bar']

    parse example #2

        my $space = Data::Object::Space->new('Foo/Bar');
      
        $space->parse;
      
        # ['Foo', 'Bar']

    parse example #3

        my $space = Data::Object::Space->new('Foo\Bar');
      
        $space->parse;
      
        # ['Foo', 'Bar']

    parse example #4

        my $space = Data::Object::Space->new('foo-bar');
      
        $space->parse;
      
        # ['FooBar']

    parse example #5

        my $space = Data::Object::Space->new('foo_bar');
      
        $space->parse;
      
        # ['FooBar']

 parts

      parts() : ArrayRef

    The parts method returns an arrayref of package namespace segments
    (parts).

    parts example #1

        my $space = Data::Object::Space->new('foo');
      
        $space->parts;
      
        # ['Foo']

    parts example #2

        my $space = Data::Object::Space->new('foo/bar');
      
        $space->parts;
      
        # ['Foo', 'Bar']

    parts example #3

        my $space = Data::Object::Space->new('foo_bar');
      
        $space->parts;
      
        # ['FooBar']

 prepend

      prepend(Str @args) : Object

    The prepend method modifies the object by prepending to the package
    namespace parts.

    prepend example #1

        # given: synopsis
      
        $space->prepend('etc');
      
        # 'Etc/Foo/Bar'

    prepend example #2

        # given: synopsis
      
        $space->prepend('etc', 'tmp');
      
        # 'Etc/Tmp/Foo/Bar'

 rebase

      rebase(Str @args) : Object

    The rebase method returns an object by prepending the package namespace
    specified to the base of the current object's namespace.

    rebase example #1

        # given: synopsis
      
        $space->rebase('zoo');
      
        # Zoo/Bar

 root

      root() : Str

    The root method returns the root package namespace segments (parts).
    Sometimes separating the root from the parts helps identify how
    subsequent child objects were derived.

    root example #1

        # given: synopsis
      
        $space->root
      
        # Foo

 routine

      routine(Str $arg1) : CodeRef

    The routine method returns the subroutine reference for the given
    subroutine name.

    routine example #1

        package Foo;
      
        sub cont {
          [@_]
        }
      
        sub abort {
          [@_]
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo');
      
        $space->routine('cont')
      
        # sub { ... }

 routines

      routines() : ArrayRef

    The routines method searches the package namespace for routines and
    returns their names.

    routines example #1

        package Foo::Routines;
      
        sub start {
          1
        }
      
        sub abort {
          1
        }
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/routines');
      
        $space->routines
      
        # ['start', 'abort']

 scalar

      scalar(Str $arg1) : Any

    The scalar method returns the value for the given package scalar
    variable name.

    scalar example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our $root = '/path/to/file';
      
        package main;
      
        $space->scalar('root')
      
        # /path/to/file

 scalars

      scalars() : ArrayRef

    The scalars method searches the package namespace for scalars and
    returns their names.

    scalars example #1

        # given: synopsis
      
        package Foo::Bar;
      
        our $root = 'root';
        our $base = 'path/to';
        our $file = 'file';
      
        package main;
      
        $space->scalars
      
        # ['root', 'base', 'file']

 sibling

      sibling(Str $arg1) : Object

    The sibling method returns a new Data::Object::Space object for the
    sibling package namespace.

    sibling example #1

        # given: synopsis
      
        $space->sibling('baz')
      
        # Foo::Baz

 siblings

      siblings() : ArrayRef[Object]

    The siblings method searches %INC and @INC and retuns a list of
    Data::Object::Space objects for each sibling namespace found (one level
    deep).

    siblings example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('encode/m_i_m_e');
      
        $space->siblings
      
        # [
        #   'Encode/Alias',
        #   'Encode/Config'
        #   ...
        # ]

 used

      used() : Str

    The used method searches %INC for the package namespace and if found
    returns the filepath and complete filepath for the loaded package,
    otherwise returns falsy with an empty string.

    used example #1

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/xyz');
      
        $space->used
      
        # ''

    used example #2

        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('c_p_a_n');
      
        $space->load;
        $space->used
      
        # 'CPAN'

    used example #3

        package Foo::Bar;
      
        sub import;
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/bar');
      
        $space->used
      
        # 'Foo/Bar'

 variables

      variables() : ArrayRef[Tuple[Str, ArrayRef]]

    The variables method searches the package namespace for variables and
    returns their names.

    variables example #1

        package Etc;
      
        our $init = 0;
        our $func = 1;
      
        our @does = (1..4);
        our %sets = (1..4);
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('etc');
      
        $space->variables
      
        # [
        #   ['arrays', ['does']],
        #   ['hashes', ['sets']],
        #   ['scalars', ['func', 'init']],
        # ]

 version

      version() : Maybe[Str]

    The version method returns the VERSION declared on the target package,
    if any.

    version example #1

        package Foo::Boo;
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/boo');
      
        $space->version
      
        # undef

    version example #2

        package Foo::Boo;
      
        our $VERSION = 0.01;
      
        package main;
      
        use Data::Object::Space;
      
        my $space = Data::Object::Space->new('foo/boo');
      
        $space->version
      
        # '0.01'

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/data-object-space/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/data-object-space/wiki>

    Project <https://github.com/iamalnewkirk/data-object-space>

    Initiatives
    <https://github.com/iamalnewkirk/data-object-space/projects>

    Milestones
    <https://github.com/iamalnewkirk/data-object-space/milestones>

    Contributing
    <https://github.com/iamalnewkirk/data-object-space/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/data-object-space/issues>



CHANGES
Changelog for Data-Object-Space

2.09 2020-07-02T15:45:31
 - Add release updates
 - Add "reload" method (#71)
 - Never attempt to load "main" package (#69)
 - Add "chain" method (#67)

2.08 2020-06-05T02:39:41
 - Built release version 2.08
 - Add release updates
 - Add "tryload" method (#62)
 - Add "all" method to exec on all parents (#63)
 - Fix bad invocant on "AUTOLOAD" call (#64)

2.07 2020-05-20T18:52:31
 - Built release version 2.07
 - Add release updates
 - Fix typo in .gitattributes
 - Add support for using "use" in a package space (#57)
 - Support reading from "DATA" token (#55)
 - Add support for using "require" in a package space (#56)
 - Address pre-perl-5.18.2 Sub/Util issue (#54)

2.06 2020-05-20T00:27:09
 - Built release version 2.06
 - Add release updates
 - Housekeeping
 - Add support for packages with autoloading
 - Add support for crude package monkey-patching
 - Updated documentation
 - Add "authority" method

2.05 2020-05-07T03:10:10
 - Built release version 2.05
 - Added release updates
 - Fix "locate" example-1 test
 - Updated documentation
 - Fix license mismatch
 - Hide shell scripts from GH stats (another way)
 - Hide shell scripts from GH stats
 - Built release version 2.04
 - Fix release process
 - Re-enable automated CPAN release

2.03 2020-05-06T02:30:19
 - Built release version 2.03
 - Added release updates
 - Add "locate" method
 - Add "rebase" method
 - Add "destroy" method
 - Rename "inc" to "included"
 - Update branching and workflow

2.02 2020-05-01T01:49:41
 - Built release version 2.02
 - Added release updates
 - Add "inc" method
 - Add ability to eval within a package
 - Add "loaded" method
 - Add meta data

2.01 2020-03-14T15:43:53
 - Built release version 2.01
 - Added release updates
 - Document inheritance and type-library
 - Method to report inheritence
 - Method to report version
 - Remove "lc" of error

2.00 2020-03-07T20:21:27
 - Built release version 2.00
 - Added release updates
 - Initial implementation
 - Initial commit


lib/Data/Object/Space.pm
package Data::Object::Space;

use 5.014;

use strict;
use warnings;
use routines;

use parent 'Data::Object::Name';

# VERSION

# METHODS

my %has;

method all($name, @args) {
  my $result = [];

  my $class = $self->class;
  for my $package ($self->package, @{$self->inherits}) {
    push @$result, [$package, $class->new($package)->$name(@args)];
  }

  return $result;
}

method append(@args) {
  my $class = $self->class;

  my $path = join '/',
    $self->path, map $class->new($_)->path, @args;

  return $class->new($path);
}

method array($name) {
  no strict 'refs';

  my $class = $self->package;

  return [@{"${class}::${name}"}];
}

method arrays() {
  no strict 'refs';

  my $class = $self->package;

  my $arrays = [
    sort grep !!@{"${class}::$_"},
    grep /^[_a-zA-Z]\w*$/, keys %{"${class}::"}
  ];

  return $arrays;
}

method authority() {

  return $self->scalar('AUTHORITY');
}

method base() {

  return $self->parse->[-1];
}

method bless($data = {}) {
  my $class = $self->load;

  return CORE::bless $data, $class;
}

method build(@args) {
  my $class = $self->load;

  return $self->call('new', $class, @args);
}

method call($func, @args) {
  my $class = $self->load;

  unless ($func) {
    require Carp;

    my $text = qq[Attempt to call undefined class method in package "$class"];

    Carp::confess $text;
  }

  my $next = $class->can($func);

  unless ($next) {
    if ($class->can('AUTOLOAD')) {
      $next = sub { no strict 'refs'; &{"${class}::${func}"}(@args) };
    }
  }

  unless ($next) {
    require Carp;

    my $text = qq[Unable to locate class method "$func" via package "$class"];

    Carp::confess $text;
  }

  @_ = @args; goto $next;
}

method chain(@steps) {
  my $result = $self;

  for my $step (@steps) {
    my ($name, @args) = (ref($step) eq 'ARRAY') ? @$step : ($step);

    $result = $result->$name(@args);
  }

  return $result;
}

method child(@args) {

  return $self->append(@args);
}

method children() {
  my %list;
  my $path;
  my $type;

  $path = quotemeta $self->path;
  $type = 'pm';

  my $regexp = qr/$path\/[^\/]+\.$type/;

  for my $item (keys %INC) {
    $list{$item}++ if $item =~ /$regexp$/;
  }

  my %seen;

  for my $dir (@INC) {
    next if $seen{$dir}++;

    my $re = quotemeta $dir;
    map { s/^$re\///; $list{$_}++ }
    grep !$list{$_}, glob "$dir/@{[$self->path]}/*.$type";
  }

  my $class = $self->class;

  return [
    map $class->new($_),
    map {s/(.*)\.$type$/$1/r}
    sort keys %list
  ];
}

method class() {

  return ref $self;
}

method cop($func, @args) {
  my $class = $self->load;

  unless ($func) {
    require Carp;

    my $text = qq[Attempt to cop undefined object method from package "$class"];

    Carp::confess $text;
  }

  my $next = $class->can($func);

  unless ($next) {
    require Carp;

    my $text = qq[Unable to locate object method "$func" via package "$class"];

    Carp::confess $text;
  }

  return sub { $next->(@args ? (@args, @_) : @_) };
}

method data() {
  no strict 'refs';

  my $class = $self->package;

  local $.;

  my $handle = \*{"${class}::DATA"};

  return '' if !fileno $handle;

  seek $handle, 0, 0;

  my $data = join '', <$handle>;

  $data =~ s/^.*\n__DATA__\r?\n/\n/s;
  $data =~ s/\n__END__\r?\n.*$/\n/s;

  return $data;
}

method destroy() {
  require Symbol;

  Symbol::delete_package($self->package);

  my $c_re = quotemeta $self->package;
  my $p_re = quotemeta $self->path;

  map {delete $has{$_}} grep /^$c_re/, keys %has;
  map {delete $INC{$_}} grep /^$p_re/, keys %INC;

  return $self;
}

method eval(@args) {
  local $@;

  my $result = eval join ' ', map "$_", "package @{[$self->package]};", @args;

  Carp::confess $@ if $@;

  return $result;
}

method functions() {
  my @functions;

  no strict 'refs';

  require Function::Parameters::Info;

  my $class = $self->package;
  for my $routine (@{$self->routines}) {
    my $code = $class->can($routine) or next;
    my $data = Function::Parameters::info($code);

    push @functions, $routine if $data && !$data->invocant;
  }

  return [sort @functions];
}

method hash($name) {
  no strict 'refs';

  my $class = $self->package;

  return {%{"${class}::${name}"}};
}

method hashes() {
  no strict 'refs';

  my $class = $self->package;

  return [
    sort grep !!%{"${class}::$_"},
    grep /^[_a-zA-Z]\w*$/, keys %{"${class}::"}
  ];
}

method id() {

  return $self->label;
}

method init() {
  my $class = $self->package;

  if ($self->routine('import')) {

    return $class;
  }

  $class = $self->locate ? $self->load : $self->package;

  if ($self->routine('import')) {

    return $class;
  }
  else {

    my $import = sub { $class };

    $self->inject('import', $import);
    $self->load;

    return $class;
  }
}

method inherits() {

  return $self->array('ISA');
}

method included() {

  return $INC{$self->format('path', '%s.pm')};
}

method inject($name, $coderef) {
  my $class = $self->package;

  local $@;
  no strict 'refs';
  no warnings 'redefine';

  if (state $subutil = eval "require Sub::Util") {
    return *{"${class}::${name}"} = Sub::Util::set_subname(
      "${class}::${name}", $coderef || sub{$class}
    );
  }
  else {
    return *{"${class}::${name}"} = $coderef || sub{$class};
  }
}

method load() {
  my $class = $self->package;

  return $class if $has{$class};

  if ($class eq "main") {
    return do { $has{$class} = 1; $class };
  }

  my $failed = !$class || $class !~ /^\w(?:[\w:']*\w)?$/;
  my $loaded;

  my $error = do {
    local $@;
    no strict 'refs';
    $loaded = !!$class->can('new');
    $loaded = !!$class->can('import') if !$loaded;
    $loaded = !!$class->can('meta') if !$loaded;
    $loaded = !!$class->can('with') if !$loaded;
    $loaded = eval "require $class; 1" if !$loaded;
    $@;
  }
  if !$failed;

  do {
    require Carp;

    my $message = $error || "cause unknown";

    Carp::confess "Error attempting to load $class: $message";
  }
  if $error
  or $failed
  or not $loaded;

  $has{$class} = 1;

  return $class;
}

method loaded() {
  my $class = $self->package;
  my $pexpr = $self->format('path', '%s.pm');

  my $is_loaded_eval = $has{$class};
  my $is_loaded_used = $INC{$pexpr};

  return ($is_loaded_eval || $is_loaded_used) ? 1 : 0;
}

method locate() {
  my $found = '';

  my $file = $self->format('path', '%s.pm');

  for my $path (@INC) {
    do { $found = "$path/$file"; last } if -f "$path/$file";
  }

  return $found;
}

method methods() {
  my @methods;

  no strict 'refs';

  require Function::Parameters::Info;

  my $class = $self->package;
  for my $routine (@{$self->routines}) {
    my $code = $class->can($routine) or next;
    my $data = Function::Parameters::info($code);

    push @methods, $routine if $data && $data->invocant;
  }

  return [sort @methods];
}

method name() {

  return $self->package;
}

method parent() {
  my @parts = @{$self->parse};

  pop @parts if @parts > 1;

  my $class = $self->class;

  return $class->new(join '/', @parts);
}

method parse() {

  return [
    map ucfirst,
    map join('', map(ucfirst, split /[-_]/)),
    split /[^-_a-zA-Z0-9.]+/,
    $self->path
  ];
}

method parts() {

  return $self->parse;
}

method prepend(@args) {
  my $class = $self->class;

  my $path = join '/',
    (map $class->new($_)->path, @args), $self->path;

  return $class->new($path);
}

method rebase(@args) {
  my $class = $self->class;

  my $path = join '/', map $class->new($_)->path, @args;

  return $class->new($self->base)->prepend($path);
}

method reload() {
  my $class = $self->package;

  delete $has{$class};

  my $path = $self->format('path', '%s.pm');

  delete $INC{$path};

  no strict 'refs';

  @{"${class}::ISA"} = ();

  return $self->load;
}

method require($target) {
  $target = "'$target'" if -f $target;

  return $self->eval("require $target");
}

method root() {

  return $self->parse->[0];
}

method routine($name) {
  no strict 'refs';

  my $class = $self->package;

  return *{"${class}::${name}"}{"CODE"};
}

method routines() {
  no strict 'refs';

  my $class = $self->package;

  return [
    sort grep *{"${class}::$_"}{"CODE"},
    grep /^[_a-zA-Z]\w*$/, keys %{"${class}::"}
  ];
}

method scalar($name) {
  no strict 'refs';

  my $class = $self->package;

  return ${"${class}::${name}"};
}

method scalars() {
  no strict 'refs';

  my $class = $self->package;

  return [
    sort grep defined ${"${class}::$_"},
    grep /^[_a-zA-Z]\w*$/, keys %{"${class}::"}
  ];
}

method sibling(@args) {

  return $self->parent->append(@args);
}

method siblings() {
  my %list;
  my $path;
  my $type;

  $path = quotemeta $self->parent->path;
  $type = 'pm';

  my $regexp = qr/$path\/[^\/]+\.$type/;

  for my $item (keys %INC) {
    $list{$item}++ if $item =~ /$regexp$/;
  }

  my %seen;

  for my $dir (@INC) {
    next if $seen{$dir}++;

    my $re = quotemeta $dir;
    map { s/^$re\///; $list{$_}++ }
    grep !$list{$_}, glob "$dir/@{[$self->path]}/*.$type";
  }

  my $class = $self->class;

  return [
    map $class->new($_),
    map {s/(.*)\.$type$/$1/r}
    sort keys %list
  ];
}

method tryload() {

  return do { local $@; eval { $self->load }; int!$@ };
}

method use($target, @params) {
  my $version;

  my $class = $self->package;

  ($target, $version) = @$target if ref $target eq 'ARRAY';

  $self->require($target);

  require Scalar::Util;

  my @statement = (
    'no strict "subs";',
    (
      Scalar::Util::looks_like_number($version)
        ? "${target}->VERSION($version);" : ()
    ),
    'sub{ my ($target, @params) = @_; $target->import(@params)}'
  );

  $self->eval(join("\n", @statement))->($target, $class, @params);

  return $self;
}

method used() {
  my $class = $self->package;
  my $path = $self->path;
  my $regexp = quotemeta $path;

  return $path if $has{$class};

  for my $item (keys %INC) {
    return $path if $item =~ /$regexp\.pm$/;
  }

  return '';
}

method variables() {

  return [map [$_, [sort @{$self->$_}]], qw(arrays hashes scalars)];
}

method version() {

  return $self->scalar('VERSION');
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.015.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Namespace Class for Perl 5",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Data-Object-Space",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Data::Object::Space",
  "PREREQ_PM" => {
    "Data::Object::Name" => "2.01",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Object::Name" => "2.01",
    "Test::Auto" => "0.12",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "2.09",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Object::Name" => "2.01",
  "Test::Auto" => "0.12",
  "routines" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
